CIPs/cip-40.md

```diff
@@ -0,0 +1,475 @@
+  keccak256(
+    keccak256("Optional<int256>(bool defined,int256 value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000000 ||
+    0x0000000000000000000000000000000000000000000000000000000000000000
+  ) ||
+  keccak256(
+    keccak256("Optional<string>(bool defined,string value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000001 ||
+    keccak256("saltvalue")
+  )
+)
+encoded = 0x1901 || domainSeparator || structHash
+```
+
+### Signatures
+
+Signatures are a useful primitive for authenticated rate limiting (e.g. allowing a higher rate limit
+to the owners of verified accounts on Celo). In order to facilitate this, the entire request struct
+SHOULD be deterministically serializeable, specifically using EIP-712.
+
+Domains that make use of signatures for authentication of requests (e.g. the `PepperedDomain`
+example below) SHOULD include a field in the domain options for specification of a signature.
+Additional fields MAY also be added, such as a nonce to prevent replaying requests.
+
+Signatures SHOULD be verified over the entire request, excluding the signature itself. Specifically,
+the message against which a signature is verified SHOULD be obtained by first setting the signature
+field value to its zero value (e.g. empty string) object, then serializing the
+`GetDomainRestrictedSigRequest` object via EIP-712.
+
+Serializing the request SHOULD use the EIP-712 domain `{ name: "ODIS Domain Restricted Signature
+Request", version: "1" }` and type defined by applying the type mapping [defined
+above](#mapping-typescript-to-eip-712-types) to the `GetDomainRestrictedSigRequest` type.
+
+### HTTP Transport
+
+When requested over HTTP, which is the primary proposed transport, the following HTTP semantics
+SHOULD be followed.
+
+If a given `name` and `version` are not recognized by an ODIS signer, it SHOULD return a `501: Not
+Implemented` error. If a given `name` and `version` are recognized, but are deprecated, it SHOULD
+return a `405: Method Not Allowed` error. Signers SHOULD keep a list of all old domain names and
+version, even if they are no longer supported.
+
+When a user's request is rejected for violating the rate limiting of a domain, the service SHOULD
+respond with `429: Too Many Requests`. If the waiting would allow for additional requests against
+the same domain, the service SHOULD include the `Retry-After` header to specify how long the client
+must wait. ODIS servers MAY apply additional rate limiting, such as by IP address. Any rate limiting
+rules outside of those applied per domain MUST be counted against first, such that a request
+dropped by optional rate limiting rules will not be counted against the domain quota.
+
+### Key management
+
+Because ODIS' current OPRF function operates with the same cryptographic primitives as the [proposed
+function](https://github.com/celo-org/celo-proposals/pull/234), and because its a bad idea in
```

Comment:
I'd recommend linking to the CIP number over the pull request.

Reply:
Oops, that's actually supposed to be a link to Notion for now, and to the crypto CIP later.